src/future/future/flatten.rs

```diff
@@ -0,0 +1,46 @@
+use futures_core::ready;
+use std::pin::Pin;
+
+use crate::future::Future;
+use crate::future::IntoFuture;
+use crate::task::{Context, Poll};
+
+#[doc(hidden)]
+#[derive(Debug)]
+pub enum FlattenFuture<Fut1, Fut2> {
+    First(Fut1),
+    Second(Fut2),
+    Empty,
+}
```

Comment:
We should be using `pin-project-lite` here to prevent the use of `unsafe` later on

Reply:
My understanding is that `pin-project-lite` does not support enums but `pin-project` does. Should I add `pin-project` to the dependencies and use it instead ?

Reply:
@taiki-e What do you suggest we do here?

Reply:
I think there are three options.
1. Use `pin_project` that supports enums. - However, it adds proc-macro related dependencies.
2. Add enum support to `pin-project-lite`. - However, the macro implementation needs to be relatively simple because complex declarative macros are slow. (When developing locally, compiled dependencies are cached, so `*-lite` crate requires both run time and initial compile time to be short.) And I'm not sure if this is possible...
3. Implement flatten using struct and Option instead of enum. - However, this is not very efficient if you think about size of flatten future.
  ```rust
   pub struct FlattenFuture<Fut1, Fut2> {
      first: Option<Fut1>,
      second: Option<Fut2>,
   }
   impl<Fut1> Future for FlattenFuture<Fut1, <Fut1::Output as IntoFuture>::Future>
   where
      Fut1: Future,
      Fut1::Output: IntoFuture,
   {
      type Output = <Fut1::Output as IntoFuture>::Output;
      fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
         let mut this = self.project();
         if let Some(fut1) = this.first.as_mut().as_pin_mut() {
             let fut2 = ready!(fut1.poll(cx)).into_future();
             this.second.set(Some(fut2));
             this.first.set(None);
         }
         if let Some(fut2) = this.second.as_mut().as_pin_mut() {
             let v = ready!(fut2.poll(cx));
             this.second.set(None);
             Poll::Ready(v)
         } else {
             unreachable!()
         }
      }
   }
  ```

Reply:
Strictly speaking it isn't correct that the last case is unreachable — it will happen if the future is polled after it has finished.

Reply:
Calls to poll on a completed future are allowed to panic and unreachable causes a panic. Not sure if the convention is to call panic explicitly.

Reply:
I would say the convention is to panic in some way that emits an error message which says a future was polled after its completion.